client-course-schedulizer/src/readCSV.ts

```diff
@@ -0,0 +1,428 @@
+const amReg = RegExp("[Aa][Mm]");
+const pmReg = RegExp("[Pp][Mm]");
+const fallReg = RegExp("[Ff]");
+const summerReg = RegExp("[Ss][Uu]");
+const springReg = RegExp("[Ss]");
+// "W" represents interim in Pruim's system it seems
+const interimReg = RegExp("[Ii]|[Ww]");
+const firstReg = RegExp("[Ff]irst");
+const secondReg = RegExp("[Ss]econd");
+const fullReg = RegExp("[Ff]ull");
+const sunReg = RegExp("[Ss][Uu]|[Nn]");
+const monReg = RegExp("[Mm]");
+const tuesReg = RegExp("[Tt](?![Hh])");
+const wedReg = RegExp("[Ww]");
+const thursReg = RegExp("[Tt][Hh]|[Rr]");
+const friReg = RegExp("[Ff]");
+const satReg = RegExp("[Ss](?![Uu])");
+
+const convertToInterface = function convertToInterface(objects: papa.ParseResult<never>) {
+  // Define variables for Schedule creation
+  let name: string;
+  let prefixes: string[];
+  let number: string;
+  let letter: string;
+  let studentHours: number;
+  let facultyHours: number;
+  let hourPart: string;
+  let numHourPart: number;
+  let minPart: string;
+  let ampm: string;
+  let startTime: string;
+  let duration: number;
+  let building: string;
+  let roomNumber: string;
+  let roomCapacity: number;
+  let year: number;
+  let term: di.Term;
+  let half: di.Half;
+  let days: di.Day[];
+  let globalMax: number;
+  let localMax: number;
+  let anticipatedSize: number;
+  let instructors: di.Instructor[];
+  let comments: string;
+  // eslint-disable-next-line prefer-const
+  const schedule: di.Schedule = {
+    sections: [],
+  };
+
+  // Get data and fields from the CSV
+  const { data, meta } = objects;
+  const { fields } = meta;
+
+  // From the CSV fields, take the ones which we recognize
+  // eslint-disable-next-line prefer-const
+  let usableFields: string[] = [];
+  let field;
+  let startTimeIdx = -1;
+  let startTimeStrIdx = -1;
+  let instructorIdx = -1;
+  let instructorsIdx = -1;
+  let prefixIdx = -1;
+  let prefixesIdx = -1;
+  if (fields) {
+    for (let i = 0; i < fields.length; i += 1) {
+      field = fields[i];
+      if (validFields.includes(field)) {
+        usableFields.push(field);
+        if (field === "startTime") {
+          startTimeIdx = usableFields.length - 1;
+        } else if (field === "startTimeStr") {
+          startTimeStrIdx = usableFields.length - 1;
+        } else if (field === "instructor") {
+          instructorIdx = usableFields.length - 1;
+        } else if (field === "instructors") {
+          instructorsIdx = usableFields.length - 1;
+        } else if (field === "prefix") {
+          prefixIdx = usableFields.length - 1;
+        } else if (field === "prefixes") {
+          prefixesIdx = usableFields.length - 1;
+        }
+      }
+    }
+  }
+
+  // If both "startTime" and "startTimeStr" are present, ignore "startTime"
+  // This is because Pruim's data has a timestamp in "startTime" (not our timezone)
+  if (startTimeIdx !== -1 && startTimeStrIdx !== -1) {
+    usableFields.splice(startTimeIdx, 1);
+  }
```

Comment:
Use `fields.forEach()`. It's also probably better to make a list of possible duplicates and not to keep track of all the indexes for splicing. That way, we can adjust the possible duplicates if the spreadsheet requirements change simply by adding to the duplicates array. In practice, I think it could look something like this
```
const usableFields: string[] = [];
const duplicates = [
    ["startTime", "startTimeStr"],
    ["prefix", "prefixes"],
    ["instructor", "instructors"],
];
if (fields) {
  fields.forEach((field) => {
      if (validFields.includes(field)) {
        usableFields.push(field);
      }
  });
}

duplicates.forEach((duplicate) => {
  if (duplicate[0] in usableFields && duplicate[1] in usableFields) {
    usableFields.splice(usableFields.indexOf(duplicate[0]), 1);
  }
});
```

Reply:
This is cool since it makes our code more flexible rather than fragile. I think you don't even need to have `if (fields)` if you are doing `fields`.

Reply:
I just realized that the first loop can be better written as
````
const usableFields = validFields.filter((field) => {
    return validFields.includes(field);
});
```

Reply:
I just realized that the first loop can be better written as
```
usableFields = fields.filter((field) => {
     return validFields.includes(field);
 });
```

Reply:
Thanks a ton for the code here, I wasn't quite sure how best to write this part, but your code is way more compact (one thing I had to change was that `duplicate[0] in usableFields` needed to be `usableFields.includes(duplicate[0])` because `in` only looks at the indexes of the array for some reason).

a413c2b